src/alert/__tests__/runtime-content.test.tsx

```diff
@@ -0,0 +1,305 @@
+    runReplacer() {
+      return {
+        update: () => {},
+        unmount: unmountCallback,
+      };
+    },
+  };
+  awsuiPlugins.alertContent.registerContentReplacer(plugin);
+  const { unmount } = render(<Alert>Alert content</Alert>);
+  await delay();
+  expect(unmountCallback).not.toBeCalled();
+  unmount();
+  expect(unmountCallback).toBeCalled();
+});
+
+test('calls update callback', async () => {
+  const callback = jest.fn();
+  const plugin: AlertFlashContentConfig = {
+    id: 'test-content',
+    runReplacer: jest.fn(() => {
+      return {
+        update: callback,
+        unmount: () => {},
+      };
+    }),
+  };
+  awsuiPlugins.alertContent.registerContentReplacer(plugin);
+  const { rerender } = render(<Alert>Alert content</Alert>);
+  await delay();
+  expect(callback).toBeCalledTimes(0);
+  expect(plugin.runReplacer).toBeCalledTimes(1);
+  rerender(<Alert>Alert new content</Alert>);
+  expect(callback).toBeCalledTimes(1);
+  expect(plugin.runReplacer).toBeCalledTimes(1);
+});
+
+describe('asynchronous rendering', () => {
+  test('allows asynchronous rendering of content', async () => {
+    const asyncContent: AlertFlashContentConfig = {
+      id: 'test-content-async',
+      runReplacer(context, registerReplacement) {
+        (async () => {
+          await pause(1000);
+          const content = document.createElement('div');
+          content.append('New content');
+          content.dataset.testid = 'test-content-async';
+          registerReplacement('content', container => {
+            container.appendChild(content);
+          });
+        })();
+        return {
+          update: () => {},
+          unmount: () => {},
+        };
+      },
+    };
+    awsuiPlugins.alertContent.registerContentReplacer(asyncContent);
+    const { container } = render(<Alert>Alert content</Alert>);
+    const alertWrapper = new AlertWrapper(container);
+    await delay();
+    expectAlertContent(alertWrapper, {
+      content: 'Alert content',
+      contentReplaced: false,
+    });
+    await delay(1000);
+    expectAlertContent(alertWrapper, {
+      content: 'New content',
+      contentReplaced: true,
+    });
+  });
+
+  test('warns if registerReplacement called after unmounting', async () => {
+    const consoleWarnSpy = jest.spyOn(console, 'warn').mockImplementation();
+    const asyncContent: AlertFlashContentConfig = {
+      id: 'test-content-async',
+      runReplacer(context, registerReplacement) {
+        (async () => {
+          await pause(1000);
+          registerReplacement('header', () => {});
+          registerReplacement('content', () => {});
+        })();
+        return {
+          update: () => {},
+          unmount: () => {},
+        };
+      },
+    };
+    awsuiPlugins.alertContent.registerContentReplacer(asyncContent);
+    const { unmount } = render(<Alert>Alert content</Alert>);
+    await delay(500);
+    unmount();
+    await delay(1000);
+    expect(consoleWarnSpy).toBeCalledWith(
```

Comment:
What is practical purpose of these warnings?

Reply:
If there needs to be some fallback when an alert disappears before registering, can we make the register functions return a boolean. If the result it false - it indicates the replacement failed because the target has been unmounted.

Reply:
it's to make it clear to consumers that they need to do cleanup on unmount. The function is anyway a no-op on our side if unmounted (the function they pass never gets called), but the warnings gives a hint that they may need to do other cleanup within their async code

Reply:
But the warning is not actionable - they might or might not see it in console and in the dev environment only. You cannot make code depend on that warning.